planner/core/rule_count_star_rewriter.go

```diff
@@ -0,0 +1,100 @@
+package core
+
+import (
+	"context"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/expression/aggregation"
+	"github.com/pingcap/tidb/parser/mysql"
+)
+
+/**
+The countStarRewriter is used to rewrite count(*) -> count(not null column)
+
+Pattern:
+LogcialAggregation
+       |
+   DataSource
+
+Optimize:
+Table
+  <k1 bool not null, k2 int null, k3 bigint not null>
+
+Case1 there are columns exists in datasource
+Query: select count(*) from table where k3=1
+CountStarRewriterRule: pick the shortest not null column exists in datasource
+Rewritten Query: select count(k3) from table where k3=1
+
+Case2 there is no columns exists in datasource
+Query: select count(*) from table
+ColumnPruningRule: pick k1 as the shortest not null column from origin table @Function.preferNotNullColumnFromTable
+                   datasource.columns: k1
+CountStarRewriterRule: rewrite count(*) -> count(k1)
+Rewritten Query: select count(k1) from table
+
+Attention:
+Since count(*) is directly translated into count(1) during grammar parsing,
+the rewritten pattern actually matches count(1)
+*/
+
+type countStarRewriter struct {
+}
+
+func (c *countStarRewriter) optimize(ctx context.Context, p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	return c.countStarRewriter(p, opt)
+}
+
+func (c *countStarRewriter) countStarRewriter(p LogicalPlan, opt *logicalOptimizeOp) (LogicalPlan, error) {
+	// match pattern agg(count(*)) -> datasource
+	if agg, ok := p.(*LogicalAggregation); ok {
+		// todo if all kinds of data source could be matched ?
+		if dataSource, ok := agg.Children()[0].(*DataSource); ok {
+			for _, aggFunc := range agg.AggFuncs {
+				if aggFunc.Name == "count" && len(aggFunc.Args) == 1 {
+					if constExpr, ok := aggFunc.Args[0].(*expression.Constant); ok {
+						if constExpr.Value.GetInt64() == 1 {
```

Comment:
What if I write the query as select count(100) from t1?

Reply:
This rule only apply on count(1). If the query is count(100), then the executor will compute count(100).

Reply:
I mean the optimization can also apply on `count(100)`. It doesn't hurt to optimize for this case.